packages/router-plugin/src/router.state.ts

```diff
@@ -95,101 +110,122 @@ export class RouterState {
   }
 
   private setUpStoreListener(): void {
-    this._store.select(RouterState).subscribe(s => {
-      this.routerState = s;
-    });
     this._store.select(RouterState.state).subscribe(() => {
       this.navigateIfNeeded();
     });
   }
 
-  private setUpStateRollbackEvents(): void {
-    this._router.events.subscribe(e => {
-      if (e instanceof RoutesRecognized) {
-        this.lastRoutesRecognized = e;
-      } else if (e instanceof GuardsCheckEnd) {
-        this.guardsCheckEnd(e.state);
-      } else if (e instanceof ResolveEnd) {
-        this.dispatchRouterDataResolved(e);
-      } else if (e instanceof NavigationCancel) {
-        this.dispatchRouterCancel(e);
-      } else if (e instanceof NavigationError) {
-        this.dispatchRouterError(e);
-      }
-    });
+  private setUpRouterEventsListener(): void {
+    this._router.events
+      .pipe(withLatestFrom(this._store.select(RouterState)))
+      .subscribe(([event, storeState]) => {
+        if (event instanceof NavigationStart) {
+          this.navigationStart(storeState);
+        } else if (event instanceof RoutesRecognized) {
+          this.routesRecognized(event);
+        } else if (event instanceof ResolveEnd) {
+          this.dispatchRouterDataResolved(event);
+        } else if (event instanceof NavigationCancel) {
+          this.dispatchRouterCancel(event);
+          this.reset();
+        } else if (event instanceof NavigationError) {
+          this.dispatchRouterError(event);
+          this.reset();
+        } else if (event instanceof NavigationEnd) {
+          this.reset();
+        }
+      });
   }
 
-  private guardsCheckEnd(routerState: RouterStateSnapshot): void {
-    this.routerStateSnapshot = this._serializer.serialize(routerState);
+  private navigationStart(storeState: RouterStateModel | null): void {
+    this.routerState = this._serializer.serialize(this._router.routerState.snapshot);
+
+    if (this.trigger !== RouterTrigger.Store) {
+      this.storeState = storeState;
+    }
+  }
+
+  private routesRecognized(event: RoutesRecognized): void {
+    this.lastRoutesRecognized = event;
+
     if (this.shouldDispatchRouterNavigation()) {
       this.dispatchRouterNavigation();
     }
   }
 
   private shouldDispatchRouterNavigation(): boolean {
-    if (!this.routerState) return true;
-    return !this.navigationTriggeredByDispatch;
+    if (!this.storeState) return true;
+    return this.trigger !== RouterTrigger.Store;
   }
 
   private navigateIfNeeded(): void {
-    if (!this.routerState || !this.routerState.state || this.dispatchTriggeredByRouter) {
+    if (
```

Comment:
```ts
private get isNoneNavigated(): boolean {
  return !this.storeState ||
      !this.storeState.state ||
      this.trigger === RouterTrigger.Router ||
      this._router.url === this.storeState.state.url
}


private navigateIfNeeded(): void {
   if (this.isNoneNavigated) {
     return;
   }

   this.trigger = RouterTrigger.Store;
   this._ngZone.run(() => this._router.navigateByUrl(this.storeState!.state!.url));
}
```

Reply:
Maybe `isNoneNavigated` could be called `canSkipNavigation`?

Reply:
Maybe `isNoneNavigated` could be called `canSkipNavigation`?

Reply:
Getters are useful when the condition is used in different places. This introduces abstraction without adding clarity + more emitted code because of `Object.defineProperty(....)`.

Reply:
True, maybe just a named variable would be a good way to go.